kubernetes/config/exec_provider.c

```diff
@@ -0,0 +1,103 @@
+#include "exec_provider.h"
+#include "kube_config_yaml.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#define ARGS_DELIM " "
+#define ENV_DELIM "="
+#define KUBECONFIG_EXEC_ARGS_BUFFER_SIZE 1024
+#define KUBECONFIG_EXEC_COMMAND_BUFFER_SIZE 1024
+#define KUBECONFIG_STRING_BUFFER_SIZE 1024
+#define KUBECONFIG_EXEC_RESULT_BUFFER_SIZE 4096
+
+int kube_exec_and_get_result(ExecCredential_t * exec_credential, const kubeconfig_property_t * exec)
+{
+    static char fname[] = "kube_exec_and_get_result()";
+    int rc = 0;
+
+    if (!exec || exec->type != KUBECONFIG_PROPERTY_TYPE_USER_EXEC) {
+        fprintf(stderr, "%s: This is not a valid kubeconfig exec configuration.\n", fname);
+        return -1;
+    }
+    if (!exec->command) {
+        fprintf(stderr, "%s: The kubeconfig exec command is not specified.\n", fname);
+        return -1;
+    }
+
+    char *env_string = NULL;
+    int env_string_length = 0;
+    if (exec->envs_count > 0 && exec->envs != NULL) {
+        for (int i = 0; i < exec->envs_count; i++) {
+            keyValuePair_t *pair = exec->envs[i];
+            env_string_length = strlen(pair->key) + strlen(ENV_DELIM) + strlen((char *) pair->value) + 1 /* 1 for the terminal of string */ ;
+            env_string = calloc(1, env_string_length);
+            if (!env_string) {
+                fprintf(stderr, "%s: Cannot allocate memory for environment variable.[%s]\n", fname, strerror(errno));
+                return -1;
+            }
+            snprintf(env_string, env_string_length, "%s=%s", pair->key, (char *) (pair->value));
+            rc = putenv(env_string);
```

Comment:
Use `setenv` instead of `putenv` for clearer/simpler memory semantics. Otherwise this will leak memory.

Reply:
OK.  The function ```setenv``` is very good to use.